src/python/pants/engine/internals/graph_test.py

```diff
@@ -624,6 +627,71 @@ def test_find_all_targets(transitive_targets_rule_runner: RuleRunner) -> None:
     assert {t.address for t in all_unexpanded} == {*expected, Address("", target_name="generator")}
 
 
+def test_label_filters(transitive_targets_rule_runner: RuleRunner) -> None:
+    transitive_targets_rule_runner.write_files(
+        {
+            "BUILD": dedent(
+                """\
+                target(name='t1')
+                target(name='t2', dependencies=[':t1'])
+                target(name='d1', dependencies=[':t1'])
+                target(name='d2', dependencies=[':t2!foo=bar'])
+                target(name='d3')
+                target(name='root', dependencies=[':d1!foo=bar', ':d2', ':d3!foo=xyzzy'])
```

Comment:
So, the fact that the labels syntax you chose ended up being so similar to the potential syntax for #13882 (which is also going to be very important in the next month or so) is really interesting.

I would be very, very good to find a way to unify these two ideas: essentially, using this syntax as representing a different flavor/variant of the target, similar to how #13882 might (i.e., depending on something as: `@scope=provided,resolve=hadoop3`).

In the case of the classpath usecase for #12794, the result could be that after filtering addresses, we drop the information from the address that gets created (because we don't actually want to create different targets/nodes in the graph: the label was only for filtering), but in other cases of #13882, we will want to _preserve_ the label as part of the `Address` (to allow targets with unique labels to actually be unique: [see](https://github.com/pantsbuild/pants/issues/13882#issuecomment-994150091)).

So essentially, use labels, but make it easy to strip labels because they've been used for filtering (for #12794) vs preserve them to force the destinations to be unique (for #13882).

cc @Eric-Arellano , @benjyw

Reply:
There is also a third type of parameter here which is currently implicit in build rules: execution and target platforms. See https://github.com/pantsbuild/pants/issues/13882#issuecomment-998419923 for some thoughts there.